### ISSUE TITLE
Add option to disable Gtk client-side decorations

### ISSUE THREAD

**Issue by Quintus** [dino/dino#320]
Hi,

I'm using i3 as my window manager, and GNOME 3's client-side decorations (CSD), which manifest themselves in this large and ugly window top bar, just look completely out of place there, not to mention that on a tiling WM the iconify and maximise buttons are dysfunctional. It would be nice to have an option to disable them and just use what the window manager provides.

Greetings
Marvin

**Comment by sydneyjd** [comment]
Hello,
 I agree with this as well. Dino blends in very well on a Gnome/Mate desktop, but when you start using it in some other DE/OS that doesn't use the Gnome 3 style decorations, it looks really out of place, and to put it bluntly, rather ugly.
 From what I understand from reading issue https://github.com/dino/dino/issues/315 you have no desire to be part of the Gnome project, and actually wish to be fully cross-platform, so why do you want the application to look like it belongs to the Gnome project, instead of letting the Window Manager chose what style bar to throw on top of the window?

Thanks!
-Sydney

**Comment by benrob0329** [comment]
Its up to your theme to make the title bar look right. This isn't a simple "don't use it", this is fundamental design.

**Comment by dos1** [comment]
@benrob0329 Not really. GTK+ often can't be made to look right, for instance when your window manager doesn't support their new protocol for CSD shadows. Unfortunately, decorations in GTK+ are really a trainwreck

When looking at Dino design-wise, it doesn't seem hard at all. Simply add a regular title bar with "Dino" caption and remove minimize/maximize/close buttons. In dialogs, remove the title from the header bar and put it in the title bar (optionally moving the Next/Cancel buttons to the bottom of the screen, removing the header bar altogether). As simple as that. 

However, as far as I know GTK+, that probably would mean a major reorganization of the UI structure, which seems too much work for too little gain. GTK+ just makes it way harder than it should be.

**Comment by ainola** [comment]
@benrob0329, This goes beyond simple theming:

https://www.reddit.com/r/kde/comments/7feg8o/question_can_i_have_shadows_in_gtk_windows_feels/dqbpd7z/

**Comment by benrob0329** [comment]
@ainola I never said it was simple theming, in fact I specifically said that *it wasn't just theming*. As for GTK issues on Kwin, yes a proper standard needs to be formed (but with Wayland on the way standardizing new X11 extensions is unlikely).

However, this isn't about arguing the problems with GTK, this is about discussing CSDs in Dino. IMHO they are a fundamental part of the design, just as they might be in say, Chrome/FF, Steam, and many other applications.

**Comment by ainola** [comment]
And to reference the following comment from #315:

> We also don't want people to conclude that we only target the GNOME desktop: Dino is intended to run on all operating systems (including Windows, MacOS and any desktop for Linux/BSD) and we do want to promote by publishing on various channels (like Windows 10 Store, Mac App Store). While this is future talk and we indeed focus on the GNOME desktop right now, this project is not about improving GNOME by giving it a decent XMPP client, but about improving XMPP by giving it a decent desktop client.

This is most certainly going to be achieved by *not* using gnome-centric paradigms. The headerbar is, to use a term used above, a trainwreck anywhere else but Gnome. If cross-platform is the desire of Dino, and they're sticking to GTK for whatever reason, then I would argue that the headerbar is not a worthy addition to the application (at present it only holds a few items that could easily be added as an actionbar in the bottom of the left pane).

The toolkit is present to give the application the functionality it needs to succeed, not the functionality to blend in to only one particular environment. The headerbar/CSD fulfils the latter situation.

**Comment by ainola** [comment]
Love the passive aggressive screenshot. It really adds. I will reiterate that CSD is not conducive to cross platform applications. One has to think of an application not as a 'GNOME' or 'kde' app, but an 'app'. Is blender, Krita, Libreoffice, et all a GNOME or KDE app? They are neither. They both use different toolkits to enable the application. GTK headerbars are built for GNOME apps, so by using them you are making a GNOME app.

To reiterate once again, GTK is perfectly fine. If you want a cross platform application, headerbars are not. There's no reason to get so passive aggressive on suggestions from your community.

**Comment by benrob0329** [comment]
>passive aggressive screenshot

I make a mockup, say it would work, but that I am unsure about the dev's having to maintain two titlebars. I was neither trying to be passive-aggressive, or imply that I was a developer :-)
I would like to point out that many MacOS and Windows apps use CSDs, so calling them "non-native" is not what I have seen from native apps.

**Comment by benrob0329** [comment]
Many "native" Windows and MacOS apps use CSDs, soo..

As for the "passive aggressive" screenshot, that was a mistake on my part.
FTR, I'm not a developer :-)

**Comment by ainola** [comment]
Thanks for clearing that up.

I disagree about native apps looking any good on windows/Mac. They just make everyone reinvent the wheel and often just exist for some corporate imposition. I do see the value of an application looking so consistent as to even keep the windowbar the same across all OSes. It gives more of a 'brand'. I would, however, continue to argue that the brand would be 'GNOME' instead of 'Dino'.

Ultimately, this decision lies in the hands of the excellent devs of this application. CSD or no, it's an application that has been desperately needed on the Linux desktop for some time. Kudos for making a simple, 'just works' XMPP client.

**Comment by benrob0329** [comment]
FTR. this is what Dino looks like with B00merang-Project's [Windows 10 theme](https://github.com/B00merang-Project/Windows-10)
![screenshot from 2018-05-07 00-13-05](https://user-images.githubusercontent.com/19180682/39684677-8a5b987e-518b-11e8-87ca-d97764e3f2fb.png)
Doesn't look too out of place to me, although it's not perfect.

**Comment by d9h20f** [comment]
@ainola Could you at least state what you actually mean? Your point seems to be that you don't dislike headerbars, you just dislike GNOME, and therefore dislike GTK headerbars. The fact that you "argue that the brand would be 'GNOME' instead of 'Dino'" is just silly. Who, beyond GNOME and other GNU/Linux users, even knows what 'GNOME' is, let alone what it looks like? More importantly, who cares?

**Comment by ainola** [comment]
I feel like I'm talking to a brick wall. The argument is not against GNOME, headerbars, or GTK, it is that *headerbars are a GNOME centric design*.

If you want to have a cross platform application, one needs to utilize toolkits in a cross platform way. Not even leaving Linux land, this bug report shows that the header bars are causing issues. In KDE the window decorations are messed up without kwin hacks. In i3 it gives double titlebars. *Even in other GTK environments like Mate or Cinnamon it will be out of place*.

Once more, *I do not dislike Dino's use of headerbars if it wants to be a GNOME application. It makes zero sense to use the headerbars if the goal is to be a cross platform application.*

**Comment by dos1** [comment]
The suspension of compositing in KWin is a feature, not an issue - up until very recently page flips under composited X used to be problematic and were often bringing performance penalty. It's problematic only in case of desktop tools written using libraries like SDL that, being game focused, request the WM to suspend composition by default, which KWin respects.

Even in MacOS and Windows lands, CSD apps are exceptions - and that's despite of both of these platforms providing an unified way to draw CSD's consistently, unlike the *nix world (which, by design, is unlikely to change). Apps like mpv or libraries like SDL don't have any decorations on Wayland thanks to this fad - they can't/don't want to use toolkit, and are unwilling to implement decorations by themselves, because that's obviously not in their scope. Both MacOS and Windows don't have these problems, making them hardly comparable.

CSD also breaks basic features like KWin's per-window rules. It's really a bad idea for any app that's not part of GNOME, unless it's heavily stylized, like Steam or Chromium. Most apps shouldn't be, and therefore should use SSD on X11 and Wayland.

**Comment by benrob0329** [comment]
>thanks to this fad

If CSDs are a fad, then its the longest running tech fad I've ever seen (seeing as how apps have been doing it since the days of Windows XP).

While they may be the exception on Windows, [judging by these screenshots](https://www.apple.com/macos/high-sierra/) I'd say that they aren't so obscure on MacOS.

>Both MacOS and Windows don't have these problems, making them hardly comparable.

Except for the fact that Dino is aiming to support these platforms, so they must be compared to some extent.

As for MPV and SDL, SDL Games are generally full screen and MPV could add some simple window decorations (like an X in the popup controls) if they wanted, although I never felt they were necessary while using just a WM (this is my own opinion ofc)

To be fair, Wayland has a lot of issues still, and GTK has a lot of issues as well. However the question is whether these are worth compromising on the aesthetic of the client. Frankly I don't think it is, but others think differently ofc. The issued on *nix will eventually work themselves out, either by mass shaming or standardization (although if they would work themselves out a little faster that would be nice).

**Comment by d9h20f** [comment]
> CSD also breaks basic features like KWin's per-window rules or global menu in unified place.

What? What do you think Ubuntu is doing?

**Comment by dos1** [comment]
@d9h20f It definitely does not use the menu button on the window decoration, like I do on Plasma.

The whole CSD issue basically plays down to GTK+ devs thinking that "just use GTK+" is a valid solution to everyone, hence things like the misused comparisons to Windows or macOS and unwillingess to interoperate with the rest of the world. Dino is the only GTK+3 app I sometimes use on my PC, its CSD is completely out of place here and causes issues. It doesn't matter that much to me personally as I use Dino only for testing and its design kinda makes sense, but please, try to put things into proper perspective here. GTK+'s CSD works only if you want to make GNOME app.

**Comment by d9h20f** [comment]
https://gitlab.gnome.org/GNOME/gtk/issues/new?issue%5Bassignee_id%5D=&issue%5Bmilestone_id%5D=

**Comment by genofire** [comment]
It would be nice if somebody give me the css classes for `~/.config/gtk-3.0/gtk.css` to disable this boarder:
![a](https://user-images.githubusercontent.com/6905586/53563148-77381800-3b53-11e9-9d04-6b6412e92a86.png)

It is really annouying that the [default one](https://wiki.archlinux.org/index.php/GTK%2B#Client-side_decorations) does not work

**Comment by megatux** [comment]
My 5 cents here. 
I use I3. You can hide minimize, maximize buttons [setting a gtk config. this way](https://wiki.archlinux.org/index.php/GTK+#Hide_CSD_buttons) because it makes no sense in this WM.
Not sure if possible but I think an app developer should _ASK_ to the WM if it supports some actions, like maximize/minimize and then draw it or not. Not sure if the actual freedesktop.org wm-spec definition covers that (maybe the  _NET_WM_STATE in
 https://specifications.freedesktop.org/wm-spec/wm-spec-1.4.html ? )
I still feel like CSD are mostly a patch to some impossibilities in SSD spec.

**Comment by vassilisw** [comment]
> Many "native" Windows and MacOS apps use CSDs, soo..
> 
> As for the "passive aggressive" screenshot, that was a mistake on my part.
> FTR, I'm not a developer :-)

NO, Mac OS always displays the **applications's menu bar at the top**, and this will not change ever.  

On the other hand GNOME's CSD **eliminated the menubar** and placed buttons inside the title bar without even specific guidelines. So you have to try all the buttons to find the function you are looking for. 
The vertical space is greater than the menu and the title bar together. 
Additionally is as inconvenient as it gets when you drag n drop a window you grabbed from a button.   
I can speak for hours on how much anti-user friendly CDS is...

**Comment by vassilisw** [comment]
I think developers must really think the benefits (including the ease of use) of the **classic window layout** (i.e. title, menu, tool, status bars) and at least support both versions (CSD, classic).   
For only **some minor** applications with **very limited** functions, CDS might be ok but not for me...

I have a complete study / paper of why CDS is the worst thing ever happened to GNOME GUI and why is against all graphical user-interface design principles. So if someone is really interested in influencing those gnome guys against CDS, please contact me..!

**Comment by radasbona** [comment]
I use Ubuntumate and installed the gtk3-nocsd package. Reboot and this is fixed for me. Same window borders as all others.
https://launchpad.net/ubuntu/+source/gtk3-nocsd

**Comment by komorebithrowsatable** [comment]
CSD takes too much space, and let's be honest - most of buttons located on CSD bar are never used. Nautilus and documents have good CSD implementation, but it's an exception.

**Comment by vassilisw** [comment]
> I use Ubuntumate and installed the gtk3-nocsd package. Reboot and this is fixed for me. Same window borders as all others.
> https://launchpad.net/ubuntu/+source/gtk3-nocsd

Yes, but this does not solve the usability problem at all.

**Comment by vassilisw** [comment]
> Nautilus and documents have good CSD implementation, but it's an exception.

When some apps are different than the rest, is just an inconsistency from the user perspective. (But if we criticise their implementation alone I might agree, yes)

**Comment by komorebithrowsatable** [comment]
It look inconsistent and ugly even in KDE where CSD protocol is implemented. Some CSD apps also doesn't respect button-layout setting.

**Comment by MorsMortium** [comment]
https://codeberg.org/MorsMortium/GTK4-NoCSD

<img width="712" height="516" alt="Image" src="https://github.com/user-attachments/assets/8ccc7dd4-90c0-4071-a5b0-34c981d1950b" />

<img width="702" height="579" alt="Image" src="https://github.com/user-attachments/assets/4d0f8eb8-c3a3-45ad-97eb-a6195de278d8" />
